session_ii/javascript-fundamentals/session2/homework/mpricu/function.js

```diff
@@ -0,0 +1,79 @@
+
+const pro1 = new Promise((resolve, reject) => {
+    // setTimeout(() => resolve(1), 1000);
+    reject(new Error("error1"));
+});
+
+const pro2 = new Promise((resolve, reject) => {
+    // setTimeout(() => resolve(2), 400);
+    reject(new Error("error2"));
+});
+
+const pro3 = new Promise((resolve, reject) => {
+    // setTimeout(() => resolve(3), 300);
+    reject(new Error("error3"));
+});
+
+const pro4 = new Promise((resolve, reject) => {
+    // setTimeout(() => resolve(4), 100);
+    reject(new Error("error4"));
+});
+
+const pro5 = new Promise((resolve, reject) => {
+    // setTimeout(() => resolve(5), 200);
+    reject(new Error("error5"));
+});
+
+promises = [pro1, pro2, pro3, pro4, pro5];
+
+const erra = [];
+
+// function to complete Promise.race in case all the promises rejected
+function allReject(timeout) {
+    return new Promise((resolve, reject) => 
+    setTimeout(() => reject('ihatepromises !'), timeout * 1000));
+}
+
+// each of the original promises have been chained with .catch method in the event of an exception 
+function f() {
```

Comment:
note that the function was supposed to receive the array of promises as an argument. You could then iterate through that array.

Reply:
Yes, you could iterate through the array of promises, but the promises are very limited because the promise is settled only once and you cannot compare the results between promises after that. That was my very first solution:

function useless() {

    return new Promise((resolve, reject) => {
        const arr1 = [];
        const arr2 = [];

        promises.forEach(async (promise, index) => {

            await promise
                .then((val) => {
                    arr1.push(promises[index]);
                })

                .catch(error => {
                    arr2.push(error.message);
                })

            if (arr1.length !== 0) {
                Promise.race(arr1)
                    .then(resu => {
                        resolve(resu);
                    })
            }

            if (arr2.length == promises.length) {
                reject(arr2);
            }
        })
    })
}

useless().then(
    cevine => console.log("quickest fulfillment value final is", cevine)
)
    .catch(error => console.log("rejection value final", error))

If you have a mix between resolved and rejected promises you have to run Promise.race to find the quickest fulfillment value( iterate second time), but second time the promise could be settled in a different way in real life, even rejected instead of resolved. 
This iteration is working just if the promises will have the same results all the time.
So I was not pleased with that solution and I searched very hard for a solution which will iterate only once through the array of promises. It is a very tricky solution but is doing the job.

Reply:
I'm not sure I fully understand, but let me know if this helps. You can observe when a promise is fulfilled or rejected and also how many of them. Also you can save the fulfillment values and/or rejections somewhere if you want. For instance you can count how many are fulfilled and how many are rejected like this:

```js
let fulfilledPromisesCount = 0;
let rejectedPromisesCount = 0;

promises.forEach((promise) => {
    promise.then(
        () => {
            fulfilledPromisesCount += 1;
        },

        () => {
            rejectedPromisesCount += 1;
        }
    )
});
```

Also you can keep flags that you can update when something happens and you can use that to make certain decisions.

Reply:
Yes, of course you will know how many promises will be resolved or rejected, but in case the number of resolved promises are greater than zero, you have to find the quickest response value. For that you need to run Promise.race (iterate second time) the array of resolved promises, but the result will not be real. Second time one of the resolved promise could resolve sooner or later or even reject.
My function (in fact is not my function, it is a solution of a very clever guy) will work every time with one iteration and without flags, does not matter if  the promises will be resolved or rejected.
There are 3 test scenarios:  number of resolved promises is zero, number of the rejected promises is zero and the third case is when you have a mix between resolved and rejected promises. You can test those cases commenting or uncommenting resolved or reject part for any promise in the array and all the time you will get the expected result.

Reply:
note you don't need to use `Promise.race` to see which one is the quickest.

```js
const delay = (time, value) => 
    new Promise((resolve) => 
        setTimeout(() => resolve(value), time));

const promises = [
    delay(1000, 4),
    delay(500, 2),
    delay(2000, 100)
];

let oneWasFulfilled = false;

promises.forEach(promise => {
    promise.then((value) => {
        if (!oneWasFulfilled) {
            console.log("the first fulilled promise value", value);

            oneWasFulfilled = true;
        }
    });
});
```

Reply:
OK, this is a trick because all your promises will be resolved. This is the difficult part: to split the array of the promises in 2, one array with the resolved promises and the other array with the rejected promises.
My function will resolve those real life scenarios when you have multiple promises with unknown settled results and you have also a timeout function which assure you get the right response no matter what.
For me those theoretical exercises are useless if I cannot find those scenarios in real life. 
For you maybe this kind of exercises are normal and you did this many times, in fact for me this situation is not frequent and I did not solved many exercises like that..

Reply:
The purpose of this exercise is to get used to the mechanism behind promises, to get used to how promises work. It doesn't have to be a real case, although there is a plan to introduce a similar operation as the one that this exercise requires in the language itself. Also, if we want to have a greater separation of concerns, which in general is a good idea, we should keep the timeout mechanism separate from other promise operations (like `Promise.all`, `Promise.race` - these are also called promise combinators). As discussed, more real scenarios will be tackled during the sessions dedicated to the frameworks. For more discussions regarding these aspects, that are not necessarily about this exercise itself, please use the Discord channel of communication.

Reply:
If you refer to the Promise.allSettled() method (which is a proposal), that will return just the 	status and the values for the settled promises. 
To get the values of the promises I found a nice solution using map:
Promise.all(promises.map(p => p.catch(e => e.message)))
    .then(results => console.log("results",results))
If we do not use my function, I still do not understand how can be determined the quickest fulfillment value with just one iteration of the promises' array.